modules/network/src/main/java/org/apache/ignite/internal/network/netty/OutboundEncoder.java

```diff
@@ -98,11 +98,17 @@ private NetworkMessageChunkedInput(
             this.serializationService = serializationService;
             this.msg = outObject.networkMessage();
 
-            List<ClassDescriptorMessage> outDescriptors = outObject.descriptors().stream()
-                    .filter(classDescriptorMessage -> !serializationService.isDescriptorSent(classDescriptorMessage.descriptorId()))
-                    .collect(Collectors.toList());
+            List<ClassDescriptorMessage> outDescriptors = null;
+            for (ClassDescriptorMessage classDescriptorMessage : outObject.descriptors()) {
```

Comment:
Is it possible to change to a counting cycle?

Reply:
It is possible, but foreach loop doesn't perform worse than the classic `for` loop on indexes. I checked allocations in the profiler, the iterator is not even allocated (probably, escape analysis does its trick). On the other hand, foreach loop is more readable and less prone to coding errors (less moving parts are in the code where we can screw up)

Reply:
I'll assume that it is being created but it's just not visible. Your ticket is about optimization, let's do it to the maximum, the counting loop in this case will make 1 line of code more, which will not lose the readability of the code.

Reply:
It is certainly not created on the heap, so what happens is allocation of a variable on the stack. The same thing that is done in an indexed for loop, but more readable and more difficult to break.

Reply:
From my experience, I can say that I came across a situation in projects when we started to see a lot of iterator allocations in the product profiler in places where this should have been optimized as well. I assume that this could happen because JIT threw out optimizations for various reasons.

Reply:
Changed to indexed for loop